File: 20 - Valid Parentheses.java
class Solution {
    public boolean isValid(String s) {
        if (s.length() % 2 != 0) {
            return false;
        }
        
        Stack<Character> stack = new Stack<>();
        for (char c: s.toCharArray()) {
            if (c == '(' || c == '[' || c == '{') {
                stack.push(c);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                char leftBracket = stack.pop();
                if (!isPair(leftBracket, c)) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
    
    private boolean isPair(char left, char right) {
        return (left == '(' && right == ')') ||
               (left == '[' && right == ']') ||
               (left == '{' && right == '}');
    }
}

File: 20 - Valid Parentheses.swift
class Solution {
    func isValid(_ s: String) -> Bool {
        let leftBrackets: [Character] = ["(", "[", "{"]
        var stack = [Character]()
        
        for c in s {
            if leftBrackets.contains(c) {
                stack.append(c)
            } else {
                guard let leftBracket = stack.last, pair(leftBracket, c) else {
                    return false
                }
                stack.removeLast()
            }
        }
        
        return stack.isEmpty
    }
    
    private func pair(_ lhs: Character, _ rhs: Character) -> Bool {
        return (lhs == "(" && rhs == ")")
            || (lhs == "[" && rhs == "]")
            || (lhs == "{" && rhs == "}")
    }
}

